src/tr2/game/level.c

```diff
@@ -31,66 +31,85 @@
+static void M_InitialiseSoundEffects(const char *file_name);
+static void M_CompleteSetup(const GF_LEVEL *level);
 
-static void M_InitialiseSoundEffects(void)
+static void M_InitialiseSoundEffects(const char *const file_name)
 {
     BENCHMARK benchmark = Benchmark_Start();
-    const char *const file_name = "data\\main.sfx";
     const char *full_path = File_GetFullPath(file_name);
     LOG_DEBUG("Loading samples from %s", full_path);
-    MYFILE *const fp = File_Open(full_path, FILE_OPEN_READ);
+    VFILE *const fp = VFile_CreateFromPath(full_path);
     Memory_FreePointer(&full_path);
 
     LEVEL_INFO *const info = Level_GetInfo();
+    VECTOR *const entries =
+        Vector_CreateAtCapacity(sizeof(WAV_ENTRY), SFX_NUMBER_OF);
+
     if (fp == nullptr) {
         Shell_ExitSystemFmt("Could not open %s file", file_name);
         goto finish;
     }
 
-    // TODO: refactor these WAVE/RIFF shenanigans
-    int32_t sample_id = 0;
-    for (int32_t i = 0; sample_id < info->samples.offset_count; i++) {
-        char header[0x2C];
-        File_ReadData(fp, header, 0x2C);
-        if (*(int32_t *)(header + 0) != 0x46464952
-            || *(int32_t *)(header + 8) != 0x45564157
-            || *(int32_t *)(header + 36) != 0x61746164) {
-            LOG_ERROR("Unexpected sample header for sample %d", i);
+    while (true) {
+        const size_t pos = VFile_GetPos(fp);
+        if (pos >= fp->size) {
+            break;
+        }
+
+        uint32_t header[11];
+        VFile_Read(fp, header, 11 * sizeof(uint32_t));
+        if (header[0] != MKTAG('R', 'I', 'F', 'F')
+            || header[2] != MKTAG('W', 'A', 'V', 'E')
+            || header[9] != MKTAG('d', 'a', 't', 'a')) {
+            LOG_ERROR("Unexpected sample header for sample %d", entries->count);
             goto finish;
         }
-        const int32_t data_size = *(int32_t *)(header + 0x28);
-        const int32_t aligned_size = (data_size + 1) & ~1;
 
-        if (info->samples.offsets[sample_id] != i) {
-            File_Seek(fp, aligned_size, FILE_SEEK_CUR);
+        const int32_t aligned_size = (header[10] + 1) & ~1;
+        const WAV_ENTRY entry = {
+            .offset = pos,
+            .size = aligned_size + 11 * sizeof(uint32_t),
+        };
+        Vector_Add(entries, (void *)&entry);
+        VFile_Skip(fp, aligned_size);
+    }
+
+    for (int32_t i = 0; i < info->samples.offset_count; i++) {
+        const int32_t offset = info->samples.offsets[i];
+        if (offset < 0 || offset >= entries->count) {
+            LOG_WARNING("Invalid sample offset %d", offset);
             continue;
         }
 
-        const size_t sample_data_size = 0x2C + aligned_size;
-        char *sample_data = Memory_Alloc(sample_data_size);
-        memcpy(sample_data, header, 0x2C);
-        File_ReadData(fp, sample_data + 0x2C, aligned_size);
+        const WAV_ENTRY entry = *(const WAV_ENTRY *)Vector_Get(entries, offset);
+        char *sample_data = Memory_Alloc(entry.size);
```

Comment:
We could prevent double allocation for each sample by exposing the VFile's internal buffer. While this might improve performance, it does create some issues, as it's not the prettiest design, and we'd still end up reading all samples, not just those necessary for the level, so the benefits are uncertain.

Reply:
Actually it's triple allocation as `Audio_Sample_LoadSingle` copies the data too :D